ufl/core/external_operator.py

```diff
@@ -0,0 +1,254 @@
+        """
+
+        ufl_operands = tuple(map(as_ufl, operands))
+        Operator.__init__(self, ufl_operands)
+
+        # Process arguments and action arguments
+        arguments = tuple((as_ufl(args), is_adjoint) for args, is_adjoint in arguments)
+        self._action_args, self._arguments = self._extract_coeffs_and_args(arguments)
+
+        # Make the coefficient associated to the external operator
+        ref_coefficient = Coefficient(function_space)
+        self._ufl_function_space = ref_coefficient.ufl_function_space()
+
+        # Checks
+        if derivatives is not None:
+            if not isinstance(derivatives, tuple):
+                raise TypeError("Expecting a tuple for derivatives and not %s" % derivatives)
+            if not len(derivatives) == len(self.ufl_operands):
+                error("Expecting a size of %s for %s" % (len(self.ufl_operands), derivatives))
+
+            self.derivatives = derivatives
+            # If we have arguments, the appropriate function space has already been set up upstream
+            if len(arguments) == 0:
+                # In this block, we construct the function space on which lives the ExternalOperator
+                # accordingly to the derivatives taken, for instance in 2d:
+                #      for V = VectorFunctionSpace(...)
+                #      if u = Function(V); e = ExternalOperator(u, function_space=V)
+                #      then, e.ufl_shape = (2,)
+                #            dedu.ufl_shape = (2,2)
+                #            ...
+                # Therefore, for 'dedu' we need to construct the new TensorElement to end up with the appropriate
+                # function space and we also need to store the VectorElement corresponding to V since it is on this
+                # function_space that we will interpolate the operands.
+                s = ref_coefficient.ufl_shape
+                for i, e in enumerate(self.derivatives):
+                    s += self.ufl_operands[i].ufl_shape * e
+                original_function_space = self._make_function_space(s, sub_element=ref_coefficient.ufl_element())
+            else:
+                original_function_space = ref_coefficient.ufl_function_space()
+        else:
+            original_function_space = ref_coefficient.ufl_function_space()
+            self.derivatives = (0,) * len(self.ufl_operands)
+
+        if coefficient is None:
+            coefficient = Coefficient(original_function_space)
+        elif not isinstance(coefficient, Coefficient):
+            raise TypeError('Expecting a Coefficient and not %s', type(coefficient))
+        self._coefficient = coefficient
+        self._original_function_space = self.coefficient.ufl_function_space()
+
+        self._extop_dependencies = [self, ]
+
+        if self.derivatives == (0,) * len(self.ufl_operands):
+            self._extop_master = self
+
+    @property
+    def coefficient(self):
+        return self._coefficient
```

Comment:
Normal UFL convention is that these are zero-argument functions, not properties.

Reply:
And indeed, you inconsistently apply that convention (compare `self.arguments()` but `self.coefficient`).